prov/rxm/src/rxm_conn.c

```diff
@@ -1379,6 +1412,92 @@ static void *rxm_conn_progress(void *arg)
 	return NULL;
 }
 
+static void *rxm_conn_atomic_progress(void *arg)
+{
+	struct rxm_ep *rxm_ep = container_of(arg, struct rxm_ep, util_ep);
+	struct rxm_fabric *rxm_fabric;
+	struct fid *fids[2];
+	struct pollfd fds[2];
+	struct rxm_msg_eq_entry *entry;
+	int poll_required;
+	int nfids;
+	int ret;
+
+	assert(rxm_ep->msg_eq);
+	rxm_fabric = container_of(rxm_ep->util_ep.domain->fabric,
+				  struct rxm_fabric, util_fabric);
+	entry = calloc(1, RXM_MSG_EQ_ENTRY_SZ);
+	if (!entry) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to allocate memory!\n");
+		return NULL;
+	}
+
+	fids[0] = &rxm_ep->msg_eq->fid;
+	ret = fi_control(fids[0], FI_GETWAIT, (void *) &fds[0].fd);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to get MSG_EP EQ WAIT_FD %d\n", ret);
+		goto exit;
+	}
+	fds[0].events = POLLIN;
+	nfids = 1;
+
+	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+	       "Starting CM conn thread with atomic AUTO_PROGRESS\n");
+
+	while (1) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+		/* CQ can be bound after progress thread is started */
+		if (OFI_UNLIKELY(nfids == 1 && rxm_ep->msg_cq)) {
+			nfids = 2;
+			fids[1] = &rxm_ep->msg_cq->fid;
+			fds[1].fd = rxm_ep->msg_cq_fd;
+			fds[1].events = POLLIN;
+		}
+		poll_required = fi_trywait(rxm_fabric->msg_fabric, fids, nfids);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+
+		if (!poll_required) {
+			fds[0].revents = 0;
+			fds[1].revents = 0;
+
+			ret = poll(fds, nfids, -1);
+			if (OFI_UNLIKELY(ret == -1)) {
+				if (errno == EINTR)
+					continue;
+				FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+					"Select error %d, closing CM thread\n",
+					errno);
+				goto exit;
+			} else if (!ret) {
+				continue;
+			}
+		}
+
+		if (poll_required || fds[0].revents & POLLIN) {
+			while (1) {
+				entry->rd = rxm_eq_read(rxm_ep, RXM_CM_ENTRY_SZ,
+							entry);
+				if (OFI_UNLIKELY(!entry->rd ||
+						 entry->rd == -FI_EAGAIN))
+					break;
+				if (entry->rd < 0 &&
+				    entry->rd != -FI_ECONNREFUSED)
+					goto exit;
+				if (rxm_conn_eq_event(rxm_ep, entry))
+					goto exit;
+				memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
+			}
+		}
+		if (nfids > 1 &&  (poll_required || fds[1].revents & POLLIN))
+			rxm_ep_progress(&rxm_ep->util_ep);
+	}
```

Comment:
We could avoid checking nfids on every iteration if we have something like this:
```
while (1) {
	if (check_if_msg_cq_is_bound())
		goto poll_eq_cq;
	poll_read_eq();
}

poll_eq_cq:
while (1) {
	poll_read_eq_cq();
}
```
What do you think?

Reply:
Yes, I think something like this makes sense. I will update the code and refresh the PR. Thanks for your comments/suggestions.

Reply:
@a-ilango - I made updates and refreshed the PR. Thanks.

Reply:
@a-ilango, @shefty -  I see an error case here that I need to fix. Do you want me to split each processing loop into separate functions (pre bind of CQ) and post bind while I do this?